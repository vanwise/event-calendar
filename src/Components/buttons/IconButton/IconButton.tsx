import styled, { css, CSSProp } from 'styled-components/macro';
import { CrossIcon, TrashIcon } from 'Components/svg';

type Icon = keyof typeof ICONS;
export interface IconButtonProps
  extends Omit<ButtonPropsWithoutRef, 'children'> {
  icon: Icon;
  iconCSS?: CSSProp;
}

const customCSS = css<{ $CSS?: CSSProp }>`
  ${({ $CSS }) => $CSS}
`;

const ICONS = {
  cross: styled(CrossIcon)`
    fill: var(--red2);
    ${customCSS}
  `,
  trash: styled(TrashIcon)`
    fill: var(--violet);
    ${customCSS}
  `,
};

function IconButton({ icon, iconCSS, ...restProps }: IconButtonProps) {
  const Icon = ICONS[icon];

  return (
    <Root type="button" {...restProps}>
      <Icon $CSS={iconCSS} />
    </Root>
  );
}

const Root = styled.button`
  width: 20px;
  transition: 0.3s ease-out;

  &:hover {
    opacity: 0.7;
  }
`;

export default IconButton;
