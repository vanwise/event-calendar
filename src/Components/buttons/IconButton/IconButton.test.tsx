import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import IconButton, { ICON_BUTTON_ICONS, IconButtonIcon } from './IconButton';

const onClick = jest.fn();

function setup(props = {}) {
  const renderResult = render(
    <IconButton icon="cross" onClick={onClick} {...props} />,
  );
  const iconButton = renderResult.getByRole('button');

  return [iconButton, renderResult] as const;
}

afterEach(jest.clearAllMocks);

describe('Icon Button', () => {
  it('should be clicked', async () => {
    const [iconButton] = setup();

    await userEvent.click(iconButton);
    expect(onClick).toBeCalledTimes(1);
  });

  it('should have different icons', () => {
    const [iconButton, { rerender }] = setup();

    for (const icon in ICON_BUTTON_ICONS) {
      rerender(<IconButton icon={icon as IconButtonIcon} />);
      expect(iconButton).toMatchSnapshot();
    }
  });
});
