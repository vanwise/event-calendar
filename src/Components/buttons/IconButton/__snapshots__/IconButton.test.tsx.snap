// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Icon Button should have different icons 1`] = `
.c1 {
  fill: var(--red2);
}

.c0 {
  width: 20px;
  -webkit-transition: 0.3s ease-out;
  transition: 0.3s ease-out;
}

.c0:hover {
  opacity: 0.7;
}

<button
  class="c0"
  type="button"
>
  <svg
    class="c1"
    viewBox="0 0 41.756 41.756"
    xml:space="preserve"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M27.948 20.878 40.291 8.536a5 5 0 1 0-7.07-7.071L20.878 13.809 8.535 1.465a5 5 0 0 0-7.07 7.071l12.342 12.342L1.465 33.22a5 5 0 1 0 7.07 7.071l12.343-12.342 12.343 12.343c.976.977 2.256 1.464 3.535 1.464s2.56-.487 3.535-1.464a5 5 0 0 0 0-7.071L27.948 20.878z"
    />
  </svg>
</button>
`;

exports[`Icon Button should have different icons 2`] = `
.c1 {
  fill: var(--violet);
}

.c0 {
  width: 20px;
  -webkit-transition: 0.3s ease-out;
  transition: 0.3s ease-out;
}

.c0:hover {
  opacity: 0.7;
}

<button
  class="c0"
  type="button"
>
  <svg
    class="c1"
    viewBox="-32 0 512 512"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M432 32H312l-9.4-18.7A24 24 0 0 0 281.1 0H166.8a23.72 23.72 0 0 0-21.4 13.3L136 32H16A16 16 0 0 0 0 48v32a16 16 0 0 0 16 16h416a16 16 0 0 0 16-16V48a16 16 0 0 0-16-16zM53.2 467a48 48 0 0 0 47.9 45h245.8a48 48 0 0 0 47.9-45L416 128H32z"
    />
  </svg>
</button>
`;
